### PR TITLE
esp32: Add support for board variants.

### PR DESCRIPTION
**Pull request by jimmo**
On the STM32 port, many of the pyboard board definitions support "variants" which enable extra features, such as threading, double-float, etc. The idea is to have a single board definition that can support different configurations.

This was recently added to the rp2 port, and used on the WEACTSTUDIO board for different flash sizes.

This PR adds this for the esp32 port, and merges the GENERIC boards down to just `GENERIC`, `GENERIC_S2`, `GENERIC_S3`, `GENERIC_C3`.

This has the following benefits:
 - There's just a single entry on the downloads page, rather than flooding the user with a whole bunch of nearly-indentical options. Then when selecting a variant the additional information in board.md can be shown to help choose.
 - It's easier to add variants than whole new board definitions. This is going to be important for the various combinations of flash and psram.

~TODO: Update tools/autobuild/build-boards.sh to query for variants and build them.~ Done. Removed the special case for stm32/pyboard and builds all variants for all boards across all ports.

_This work was funded through GitHub Sponsors._

### CODE REVIEW

**Comment by jimmo** [comment]
Added the change to autobuild.sh to build all variants for all boards.

**Comment by jimmo** [comment]
@dpgeorge This will require a small fixup on the build server to rename exisiting firmware files to the new names. But otherwise it should allow us to keep providing the full history for these "new" esp32 variants.

**Comment by Flipje1955** [comment]
Hi Jimmo, respectfully:  the statement "This was recently added to the rp2 port, and used on the WEACTSTUDIO board for different flash sizes." may have been implemented as an approach but it to date not followed through on, as the WEACTSTUDIO download page shows.

**Comment by jimmo** [comment]
> Hi Jimmo, respectfully: the statement "This was recently added to the rp2 port, and used on the WEACTSTUDIO board for different flash sizes." may have been implemented as an approach but it to date not followed through on, as the WEACTSTUDIO download page shows.

hi @Flipje1955 -- this is exactly the purpose of the final commit in this PR. All variants for all boards in all ports will now be published to the downloads page. This includes the WEACTSTUDIO variants.

**Comment by mattytrentini** [comment]
This is looking great @jimmo! It's a _very_ handy feature. 

However, don't we need additional variants defined for the GENERIC boards, in particular the S3? 4, 8, 16 and 32MB are all common configurations, I think...

**Comment by jimmo** [comment]
> However, don't we need additional variants defined for the GENERIC boards, in particular the S3? 4, 8, 16 and 32MB are all common configurations, I think...

@mattytrentini Yep for sure. I just wanted to migrate the existing boards over as a first step.

**Comment by jimmo** [comment]
After discussing with @dpgeorge we've decided to get rid of the `query-variants` target, and just use board.json to identify the variants that should be built. No point duplicating the list of variants in both `mpconfigboard.{mk,cmake}` as well as `board.json`.

So I've removed the definition of `BOARD_VARIANTS` from all boards that currently set it. I've kept the fix for `make submodules` for ESP32 in the PR.

Also rebased after #12081

**Comment by mattytrentini** [comment]
First build I've tried appears to work:
```bash
> docker run -ti --rm -v $(pwd):/code -w /code espressif/idf:v5.0.2 bash -c "make -C mpy-cross && make -C ports/esp32 submodules all BOARD=GENERIC_S3 BOARD_VARIANT=spiram-oct"
...[snip]...
Project build complete. To flash, run this command:
/opt/esp/python_env/idf5.0_py3.8_env/bin/python ../../../opt/esp/idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after no_reset --chip esp32s3  write_flash --flash_mode dio --flash_size 8MB --flash_freq 80m 0x0 build-GENERIC_S3/bootloader/bootloader.bin 0x8000 build-GENERIC_S3/partition_table/partition-table.bin 0x10000 build-GENERIC_S3/micropython.bin
or run 'idf.py -p (PORT) flash'
bootloader  @0x000000    18288  (   14480 remaining)
partitions  @0x008000     3072  (    1024 remaining)
application @0x010000  1492224  (  539392 remaining)
total                  1557760
make: Leaving directory '/code/ports/esp32'
```

Looks _very_ promising!

**Comment by jimmo** [comment]
> @dpgeorge This will require a small fixup on the build server to rename exisiting firmware files to the new names. But otherwise it should allow us to keep providing the full history for these "new" esp32 variants.

@dpgeorge For the `d2wd`, `ota`, `unicore` variants of GENERIC, no renaming necessary as the old boards had the exact same ID set in their `board.json` as the variant file will use. So just the `spiram` variant needs to be renamed from `esp32spiram`.

```bash
for f in esp32spiram-*.*; do echo mv $f ${f/esp32spiram/esp32-spiram}; done
```

Same for the `usb` variant of GENERIC_C3, nothing to be done.

For `GENERIC_S2`, the `ESP32_S2_WROVER` board is now the `wrover` variant:

```
for f in ESP32_S2_WROVER-*.*; do echo mv $f ${f/ESP32_S2_WROVER/GENERIC_S2-wrover}; done
```

For `GENERIC_S3`, the `GENERIC_S3_SPIRAM` and `GENERIC_S3_SPIRAM_OCT` boards are now the `spiram` and `spiram-oct` variants respectively:

```
for f in GENERIC_S3_SPIRAM_OCT-*.*; do echo mv $f ${f/GENERIC_S3_SPIRAM_OCT/GENERIC_S3-spiram-oct}; done
for f in GENERIC_S3_SPIRAM-*.*; do echo mv $f ${f/GENERIC_S3_SPIRAM/GENERIC_S3-spiram}; done
```

File: extmod/extmod.cmake [hunk 1]
```diff
@@ -60,51 +60,62 @@ set(MICROPY_SOURCE_EXTMOD
...
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_overflow.c
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_page.c
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_put.c
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_search.c
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_seq.c
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_split.c
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_utils.c
-        ${MICROPY_LIB_BERKELEY_DIR}/mpool/mpool.c
-    )
+    if(ECHO_SUBMODULES OR ECHO_QUERY_VARIANTS)
```
Comment (dpgeorge) [inline extmod/extmod.cmake:65]: ECHO_QUERY_VARIANTS is not needed

File: ports/esp32/boards/GENERIC/mpconfigboard.cmake [hunk 2]
```diff
@@ -2,3 +2,50 @@ set(SDKCONFIG_DEFAULTS
...
+        boards/GENERIC/sdkconfig.unicore
+    )
+
+    list(APPEND MICROPY_DEF_BOARD
+        MICROPY_HW_BOARD_NAME="Generic ESP32 unicore module"
+    )
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "d2wd ota spiram unicore")
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC/mpconfigboard.cmake:51]: not needed

File: ports/esp32/boards/GENERIC/mpconfigboard.h [hunk 3]
```diff
@@ -1,2 +1,6 @@
-#define MICROPY_HW_BOARD_NAME "ESP32 module"
+#ifndef MICROPY_HW_BOARD_NAME
+// Can be set by mpconfigboard.cmake.
+#define MICROPY_HW_BOARD_NAME "Generic ESP32 module"
+#endif
+
 #define MICROPY_HW_MCU_NAME "ESP32"
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC/mpconfigboard.h:6]: Should this be configured per-variant?  It previously did have different values for the variants, but arguably it should be the same for all of them, so probably OK to keep it at just a single value like you've done.

File: ports/esp32/boards/GENERIC/mpconfigboard.cmake [hunk 4]
```diff
@@ -2,3 +2,50 @@ set(SDKCONFIG_DEFAULTS
     boards/sdkconfig.base
     boards/sdkconfig.ble
 )
+
+if(MICROPY_BOARD_VARIANT STREQUAL "d2wd")
+    set(SDKCONFIG_DEFAULTS
+        ${SDKCONFIG_DEFAULTS}
+        boards/GENERIC/sdkconfig.d2wd
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC/mpconfigboard.cmake:9]: this file is empty for some reason

File: ports/esp32/boards/GENERIC_C3/mpconfigboard.cmake [hunk 5]
```diff
@@ -4,3 +4,13 @@ set(SDKCONFIG_DEFAULTS
...
+
+if(MICROPY_BOARD_VARIANT STREQUAL "usb")
+    set(SDKCONFIG_DEFAULTS
+        ${SDKCONFIG_DEFAULTS}
+        boards/GENERIC_C3/sdkconfig.c3usb
+    )
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "usb")
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_C3/mpconfigboard.cmake:16]: not needed

File: ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb [hunk 6]
```diff
@@ -1,4 +1,3 @@
-CONFIG_ESP32C3_REV_MIN_3=y
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb:1]: why is this line removed?

File: ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake [hunk 7]
```diff
@@ -4,3 +4,25 @@ set(SDKCONFIG_DEFAULTS
...
+    )
+else()
+    # Enable UART REPL for modules that have an external USB-UART and don't use native USB.
+    list(APPEND MICROPY_DEF_BOARD
+        MICROPY_HW_ENABLE_UART_REPL=1
+    )
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "wrover")
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:28]: not needed

File: ports/esp32/boards/GENERIC_S3/mpconfigboard.cmake [hunk 8]
```diff
@@ -6,3 +6,29 @@ set(SDKCONFIG_DEFAULTS
...
+        boards/sdkconfig.spiram_oct
+    )
+
+    list(APPEND MICROPY_DEF_BOARD
+        MICROPY_HW_BOARD_NAME="Generic ESP32S3 module with Octal-SPIRAM"
+    )
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "spiram spiram-oct")
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_S3/mpconfigboard.cmake:34]: not needed

File: py/mkrules.cmake [hunk 9]
```diff
@@ -178,7 +178,12 @@ if(MICROPY_FROZEN_MANIFEST)
         set(MICROPY_LIB_DIR ${MICROPY_DIR}/lib/micropython-lib)
     endif()
 
-    if(NOT (${ECHO_SUBMODULES}) AND NOT EXISTS ${MICROPY_LIB_DIR}/README.md)
+    if(ECHO_SUBMODULES OR ECHO_QUERY_VARIANTS)
```
Comment (dpgeorge) [inline py/mkrules.cmake:181]: ECHO_QUERY_VARIANTS not needed

File: ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake [hunk 10]
```diff
@@ -4,3 +4,25 @@ set(SDKCONFIG_DEFAULTS
...
+if(MICROPY_BOARD_VARIANT STREQUAL "wrover")
+    set(SDKCONFIG_DEFAULTS
+        ${SDKCONFIG_DEFAULTS}
+        boards/sdkconfig.spiram_sx
+        boards/GENERIC_S2/sdkconfig.wrover
+    )
+
+    list(APPEND MICROPY_DEF_BOARD
+        MICROPY_HW_I2C0_SCL=7
+        MICROPY_HW_I2C0_SDA=6
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:17]: Did you consider putting settings like this in `mpconfigboard.h`?  Eg:
```c
#if MICROPY_BOARD_VARIANT_WROVER
#define MICROPY_HW_I2C0_SCL 7
#endif
```
Probably it's cleaner to keep all variant config in the .cmake file (and .mk for make-based ports)...

File: ports/esp32/boards/GENERIC_S2/board.json [hunk 11]
```diff
@@ -14,5 +14,8 @@
     "product": "ESP32-S2",
     "thumbnail": "",
     "url": "https://www.espressif.com/en/products/modules",
-    "vendor": "Espressif"
+    "vendor": "Espressif",
+    "variants": {
+        "wrover": "WROVER module"
+    }
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_S2/board.json:20]: Looks like this change is in the wrong commit (it's in the GENERIC_S3 commit).

File: ports/esp32/boards/GENERIC_S3_SPIRAM_OCT/mpconfigboard.cmake [hunk 12]
```diff
@@ -1,11 +0,0 @@
-set(IDF_TARGET esp32s3)
-
-set(SDKCONFIG_DEFAULTS
-    boards/sdkconfig.base
-    boards/sdkconfig.usb
-    boards/sdkconfig.ble
-    boards/sdkconfig.240mhz
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_S3_SPIRAM_OCT/mpconfigboard.cmake:7]: this 240mhz file seems to be missing in the new variant

File: tools/autobuild/build-boards.sh [hunk 13]
```diff
@@ -3,8 +3,36 @@
...
+#   $ cd ports/rp2
+#   $ MICROPY_AUTOBUILD_MAKE="make -j8" build_rp2_boards -latest /tmp
+#
+# Or to build a single board:
+#
+#   $ source tools/autobuild/build-boards.sh
+#   $ cd ports/rp2
+#   $ MICROPY_AUTOBUILD_MAKE="make -j8" build_board boards/PICO/board.json -latest /tmp uf2
+
+function copy_artifacts {
```
Comment (dpgeorge) [inline tools/autobuild/build-boards.sh:16]: The British spelling is "artefacts" :)

File: tools/autobuild/build-boards.sh [hunk 14]
```diff
@@ -13,33 +41,29 @@ function build_board {
...
-            fi
-        done
-    )
+    local build_dir=/tmp/micropython-build-$board
+    $MICROPY_AUTOBUILD_MAKE BOARD=$board BUILD=$build_dir && copy_artifacts $dest_dir $descr $fw_tag $build_dir $@
     rm -rf $build_dir
+
+    # Query variants from board.json and build them. Ignore the special "idf3"
+    # variant for ESP32 boards (this allows the downloads page to still the
+    # idf3 files for older releases that used to be explicitly built).
```
Comment (dpgeorge) [inline tools/autobuild/build-boards.sh:60]: minor grammatical error, "to still have the"

File: ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb (hunk 6 above)
Comment (jimmo) [reply ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb]: The simple answer is because it's the default anyway (without a bit of stuffing around you cannot compile firmware for revisions before 3).

I tried to piece the history together but I think what happened was that the engineering samples (rev 1 & 2) didn't support the USB peripheral, but our original GENERIC_C3 board definition was added then, and then later the GENERIC_C3_USB board was added to support the production boards.

Then later (IDF 5.0 I think, but haven't confirmed) the IDF made revision 3 the default minimum. i.e. neither GENERIC_C3 and GENERIC_C3_USB builds work by default any more on engineering samples. (I know this because I _only_ have engineering sample c3 boards).

So the longer answer is that I think we should actually switch this around and make usb the default (or possibly just not have a variant at all).

File: ports/esp32/boards/GENERIC/mpconfigboard.h (hunk 3 above)
Comment (jimmo) [reply ports/esp32/boards/GENERIC/mpconfigboard.h:6]: Yeah, so the d2wd and unicore boards did have different names before.

I think you could make the case that d2wd and unicore are attributes of the MCU whereas spiram and ota are attributes of the board/configuration. I think we should fix it.

In theory you could have e.g. a spiram+unicore variant, in which case I think setting the MCU to "esp32-unicore" and the board to "Generic ESP32 module with SPIRAM" makes sense.

Updated.

**Comment by jimmo** [comment]
Updated with review comments fixed.

File: ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb (hunk 6 above)
Comment (dpgeorge) [reply ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb]: Hmm, OK.  So maybe we should follow IDF 5 and have the default GENERIC_C3 variant be minimum rev 3 and enable USB.  Then we could have a variant which runs on < rev 3 without USB.  Is that possible/easy?  If not easy then we can just have no variants for GENERIC_C3 and enable USB by default (until someone asks for < rev 3 support).

Said differently: if we only support rev 3, then we should enable USB by default.

Comment (jimmo) [reply ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb]: Yes, I agree.

Will update later tonight.

File: ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake (hunk 10 above)
Comment (jimmo) [reply ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:17]: I did but there was no existing precedent for having the variant as a preprocessor macro, and exactly like you say it's neater putting it in one place.

Comment (jimmo) [reply ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:17]: I'm pretty confused by this "wrover" variant anyway and why it needs to define the pins?

Comment (dpgeorge) [reply ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:17]: Maybe because USB uses GPIO19 which is the default for I2C0 SDA?  But that would be a problem for all S2 boards.

Comment (jimmo) [reply ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:17]: How do you feel about replacing this variant with a "spiram" variant, which is consistent with what we've done for OG and S3?

File: ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb (hunk 6 above)
Comment (jimmo) [reply ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb]: Done.

File: ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake (hunk 10 above)
Comment (jimmo) [reply ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake]: I've pushed the branch with an extra commit that does this rename/replace. See the commit description for notes.

**Comment by jimmo** [comment]
Updated based on further discussion.

For everything except OG, there is no longer a "spiram" variant. For simplicity, we just support spiram on the default configuration and it auto-detects at boot.      (The OG needs to maintain a spiram variant because enabling spiram comes at additional costs to work around silicon bugs that you don't want to pay for if you don't have spiram).

Added a board.md for all the generic boards that explains the different variants and how they map to the Espressif module names.

Squashed the S2_WROVER commit into the main S2 commit.

Renamed the default `partitions.csv` to `partitions-4MiB.csv` to match the others.

File: ports/esp32/boards/GENERIC_S2/mpconfigboard.h [hunk 15]
```diff
@@ -1,8 +1,5 @@
-#define MICROPY_HW_BOARD_NAME               "ESP32S2 module"
-#define MICROPY_HW_MCU_NAME                 "ESP32S2"
+#define MICROPY_HW_BOARD_NAME "Generic ESP32S2 module"
+#define MICROPY_HW_MCU_NAME "ESP32S2"
 
-#define MICROPY_PY_BLUETOOTH                (0)
-#define MICROPY_HW_ENABLE_SDCARD            (0)
-
-// Enable UART REPL for modules that have an external USB-UART and don't use native USB.
-#define MICROPY_HW_ENABLE_UART_REPL         (1)
```
Comment (dpgeorge) [inline ports/esp32/boards/GENERIC_S2/mpconfigboard.h:8]: I see you removed this... what was the reason?

Related to this, it seems that `CONFIG_USB_AND_UART` is not a real thing, at least I could not find it in the IDF or TinyUSB component...

Comment (jimmo) [reply ports/esp32/boards/GENERIC_S2/mpconfigboard.h:8]: Yep, that should be still there. Updated

Also removed `CONFIG_USB_AND_UART` from all S2 boards.

**Comment by dpgeorge** [comment]
Merged, thanks for all the work on this one!

**Comment by robert-hh** [comment]
@jimmo I have a question about using this new board variant scheme. If there a port has a basic configuration and variants, does the basic configuration has to be referred in boards.json as well?

**Comment by jimmo** [comment]
@robert-hh Nope it doesn't.
